---
title: "Variables Description" 
author: "Anusha Kumar, Kyla Finlayson, Nataliya Kyrychenko"
date: "2022-10-18"
output:
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE)

#Load packages
library(tidyverse)
library(kableExtra)
library(gridExtra)
library(tableone)
library(lubridate)
library(gridExtra)

#Read in data
isolates <- read.csv("~/Desktop/isolates.csv", na.strings=c("","NA"))


```

### **Data Description**

In this study, secondary data is used for Campylobacter jejuni, taken from the "Isolates" browser of the National Library of Medicine's (NIH) NCBI Pathogen Detection Project.

This public repository is widely used by public health practitioners and researchers, both at international, federal, and state levels. It helps to match foodborne pathogens that became source of clinical cases around the US, and even the whole world through clustering of related isolates. It also helps to investigate outbreaks, research prevalence of strains, AMR and virulence genes. There are other types of use of NCBI Pathogen Detection database both by public health practitioners and researchers. For example, the database is used for international outbreaks investigation, connecting clinical and environmental isolates in Canada, USA, France, Australia, and South Korea. Except outbreak detection, the database is used for analysis of distribution of pathogens throughout the world, including specific virulent genes, AMR-resistant genes, analysis of prevalence of certain genomic sequence.

For our project, we will be using a specific subset of the databse which contains isolates of Campylobacter jejuni. Our initial dataset has 80095 observations, 19 variables, and can be located at:
https://www.ncbi.nlm.nih.gov/pathogens/isolates/#taxgroup_name:%22Campylobacter%20jejuni%22. The default dataset has 16 main variables, with the option to add other variables. We have decided to add three variables: 'Collection.date', "Outbreak', and 'Host disease' as we believe these could potentially be useful in our analysis.

### **Variable Descriptions**

First, we will begin by looking at the data definitions for all of our variables. Then, we will decide which variables we would like to keep for our analysis.

* **X.Organism.group**

"X.Organism.group" is a categorical character variable, with only one level namely "Campylobacter jejuni." This variable shows the group in which an organism is classified into by taxonomy. In this case, we have all observations as Campylobacter jejuni. As a result, we will not use this variable in our final dataset because it does not tell us any useful information.

*  **Strain**

```{r}
# structure of data by strain
str(data$Strain)
```

"Strain" is a categorical character variable. A strain can be described as a subtype or genetic variation of a microorganism (e.g., a virus, bacterium or fungus, in other words, a microbial strain name, if provided by the person who submits this information). For example, when we refer to a "flu strain," this is a specific biological variation of the influenza virus. As a result, this variable may be useful in our analysis.

* **Isolate.identifiers**
```{r}
# structure of data by isolate identifiers
str(data$Isolate.identifiers)
```

"Isolate.identifiers" is a categorical character variable. It shows the identity of each isolate, and can contain special characters (hyphens). Isolation is the process of removing a strain from a naturally occurring, diverse population of living bacteria that is present in the environment. Once a strain has undergone isolation, it becomes an "isolate". (Source: https://www.tmcc.edu/microbiology-resource-center/lab-protocols/bacterial-isolation)

* **Serovar**
```{r}
# generate serovar counts
# name as summary
Summary<-data %>% count(Serovar)
# generate names of summary
# serovar and count columns
names(Summary)<-c("Serovar","Count")
# add percentage column
# create into a table
Summary$Percentage <- round(Summary$Count/sum(Summary$Count)*100,2)

Summary %>%
  arrange(desc(Count)) %>%
  head(10) %>%
  kbl(caption = "Distribution of Serovar") %>%
  kable_classic(full_width = F, html_font = "Cambria")
```

"Serovar" is a categorical character variable with ten levels and several missing values. It represents the combined field of sub-species, serotype, or serovar, if provided. There are 79,663 missing values (99.5% of observations) with the largest type being unknown and other types labelled "not available", "not known", "not collected", etc. Out of approximately 80,000 observations, we only have about 5 observations where the actual value for Serovar is known. As a result, this variable will not be very useful for our analysis and we will likely remove it from our final dataset.

* **Isolate**

```{r, eval=FALSE}
# display structure of isolate
str(data$Isolate)
# display total count of isolate missing values
sum(is.na(data$Isolate))
```

"Isolate" is a categorical character variable that shows separate microorganisms based on microorganism's culture or subpopulation from its parent population. This variable demonstrates how to carry out this separation and control, for instance, by isolating a pathogen from sick plant tissue. This variable has no missing values and will be useful in our analysis.

* **Create.date**
```{r, eval=FALSE}
# display structure of create date
str(data$Create.date)
# generate sum of missing value counts in create date
sum(is.na(data$Create.date))
```

"Create.date" is a time variable that represents the date that each isolate was entered or created in the database. In other words, this represents the date on which the particular isolate was seen/detected by the Pathogen Detection System. The format used is the "YYYY-MM-DD T0 HR:MIN:SEC" format. It has no missing values.

* **Location**

```{r}
# display location count
# name as summary
Summary<-data %>% count(Location)
# display names of summary
# location and count columns
names(Summary)<-c("Location","Count")
# add percentage column
# form into a table
Summary$Percentage <- round(Summary$Count / sum(Summary$Count)*100,2)

Summary %>%
  arrange(desc(Count)) %>%
  head(10) %>%
  kbl(caption = "Distribution of Location") %>%
  kable_classic(full_width = F, html_font = "Cambria")
```

"Location" is a categorical character variable that represents the location where the data was collected (geographic location of the sample). There are two parts to the location, indicated by the colon (:). The name before the colon represents the country, while the name after the colon represents the region. However, a large percentage of values only have data for country, not country and state. This variable has 448 different categories (locations) listed and has 4906 missing observations. It will be very useful in our analysis because we will narrow our dataset down to only include isolates found in the United States. Furthermore, it will be useful in determining outbreaks.

* **Isolation.source**

```{r}
# add count of isolation source
Summary<-data %>% count(Isolation.source)
# generate names
# isolation source and count columns
names(Summary)<-c("Isolation.source","Count")
# add percentage 
# display in table
Summary$Percentage <- round(Summary$Count/sum(Summary$Count)*100,2)

Summary %>%
  arrange(desc(Count)) %>%
  head(10) %>%
  kbl(caption = "Distribution of Isolation.source") %>%
  kable_classic(full_width = F, html_font = "Cambria")
```

"Isolation.source" is a categorical character variable that represents the source of each isolate (where the isolate was contracted from). For instance, the source of isolation can be stool or chicken carcass among many others. More particularly, isolation source describes the physical, environmental, and/or local geographical source. This variable has 658 different categories (types of isolation sources) with approximately 18,000 missing values.

* **Isolation.type**

```{r, fig.height = 3, fig.width = 7, fig.cap = "Distribution of Isolation.type"}
# add count of isolation type
Summary<-data %>% count(Isolation.type)
# generate names
# isolation type and count columns
names(Summary)<-c("Isolation.type","Count")
# add percentage column
# display in table 
Summary$Percentage<-Summary$Count/sum(Summary$Count)*100
# create summary with distribution of isolation type
Summary%>%
  kbl(caption = "Distribution of Isolation.type") %>%
  kable_classic(full_width = F, html_font = "Cambria")
# create plot
# percentage of isolation types 
ggplot(Summary, aes(x = Isolation.type, y=Percentage,fill=Isolation.type)) +
  geom_bar(stat="identity",position=position_dodge(),width = 0.98)+theme_minimal()+ ggtitle("") +
  xlab("Isolation.type") + ylab("Percentage")+
theme(axis.text.x=element_blank(),
      axis.ticks.x=element_blank(),
      )
```

"Isolation.type" is a categorical character variable that represents the type of isolation in each observation (clinical OR environmental/other). When an isolate is observed in a human/clinical case, the isolation type will be clinical. When an isolate is observed in an environmental setting, such as a water sample, it will be environmental/other. This variable has only two categories and has approximately 3,000 missing values Environmental/other isolation types appears most in the data set, with 43282 recordings, which is approximately 54.03%.

* **SNP.cluster**

```{r, eval=FALSE}
# structure of SNP cluster
str(data$SNP.cluster)
# sum of missing value counts in SNP cluster
sum(is.na(data$SNP.cluster))
```

"SNP.cluster" is a categorical character variable that represents single nucleotide polymorphisms (SNPs), or DNA sequence variations. SNPs occur when a single nucleotide in the genome varies between matched chromosomes. More specifically, the SNP cluster is a group of isolates, where the genome assemblies are closely linked to each other. This is dependent on the type of clustering methodology.
(Source: https://medlineplus.gov/genetics/understanding/genomicresearch/snp/#:~:text=Single%20nucleotide%20polymorphisms%2C%20frequently%20called,building%20block%2C%20called%20a%20nucleotide.). In our data, this variable has 5,782 different categories (SNPs) and approximately 14,000 missing values. 

* **BioSample**

```{r, eval=FALSE}
# display structure of biosample
str(data$BioSample)
# display total count of biosample missing values
sum(is.na(data$BioSample))
```

According to the NCBI documentation, biosample records describe the biological source materials used in experimental assays. It contains descriptive information about the physical sample from which an isolate is derived (for example, a tissue biopsy). In our dataset, "Biosample" represents the accession number that each Biosample has been assigned. It is a categorical character variable with no missing values. We expect almost all values to be unique unless there are two strains collected from the same biosample. This is true, as we see "Biosample" has 80,105 unique categories which is approximately the size of the dataset.

* **AMR.genotypes**
```{r, eval=FALSE}
# structure of AMR genotypes
str(data$AMR.genotypes)
# total count of AMR genotypes missing values
sum(is.na(data$AMR.genotypes))
```

"AMR.genotypes" is a categorical character variable that represents the antimicrobial resistance (AMR) genesfound in each isolate. These AMR genes are linked to or implicated in the resistance to one or more antibiotics. The existence or absence of a gene in the same or a similar strain may be due to the acquisition of particular mutations. (Source: https://www.who.int/news-room/fact-sheets/detail/antimicrobial-resistance). This variable has 7,000 missing values and approximately 2,962 different categories. It may not be useful to our analysis, since our primary research question is not related to antibiotic resistance, however it may be a variable we find useful to adjust for so we will likely keep it in the dataset.

* **Computed.types**

```{r, eval=FALSE}
# structure of computed types
str(data$Computed.types)
# count missing values
sum(is.na(data$Computed.types))
```

"Computed.type" is a categorical character variable. represents the "computational prediction based on the sequence calculated as part of the Pathogen Detection Pipeline". Unfortunately, all of our values our missing for this variable because it is only available for Salmonella isolates. As a result, we will remove this variable from our dataset.

* **Min.same + Min.diff**
```{r, fig.height = 3, fig.width = 7, fig.cap = "Distribution of Min.same and Min.diff"}
# histogram of min same
ggplot(data, aes(x = as.numeric(Min.same))) +
  geom_bar() + 
  labs(x = "Min.same", y = "Count")

ggplot(data, aes(x = as.numeric(Min.diff))) +
  geom_bar() + 
  labs(x = "Min.diff", y = "Count")
```


"Min.same" is a continuous variable that represents the minimum SNP distance from one isolate to another of the same isolation type. The distribution of this variable seems to be skewed right, indicating a violation of the normality distribution. From analyzing the distribution of each recorded number, a distance of zero is the most common with 10,842 counts. There are a total of 33,390 missing values and the values range from 0 to 98.

"Min.diff" is a continuous variable that represents the minimum SNP distance from one isolate to another of a different isolation type. The distribution of this variable also seems to be skewed right, but less right-skewed than Min.same, still indicating a violation of the normality distribution. From analyzing the distribution of each recorded number, a distance of 2 is the most common with 3240 counts. There are 33,390 missing values and the values range from 0 to 115.

* **Collection.date**

```{r, eval=FALSE}
# display structure of collection date
str(data$Collection.date)
# generate sum of missing value counts in collection date
sum(is.na(data$Collection.date))
```

"Collection.date" is a time variable that represents the date that each isolate was collected from its respective sample. The format used is the YYYY-MM-DD format (follows ISO format). It has approximately 14,000 missing values. Furthermore, there are a large number of variables that only have values for year, or only for month and year. There are also a handful of observations that for an unknown reason list two years (for example, "1900/1971" is entered).

* **Host.disease**
```{r, eval=FALSE}
# display structure of host disease
str(data$Host.disease)
# generate sum of missing value counts in host disease
sum(is.na(data$Host.disease))
```

```{r}
# add count of isolation source
Summary<-data %>% count(Host.disease)
# generate names
# isolation source and count columns
names(Summary)<-c("Host.disease","Count")
# add percentage 
# display in table
Summary$Percentage<- round(Summary$Count / sum(Summary$Count)*100,2)

Summary %>%
  arrange(desc(Count)) %>%
  head(10) %>%
  kbl(caption = "Distribution of Host.disease") %>%
  kable_classic(full_width = F, html_font = "Cambria")
```

"Host.disease" is a categorical character variable that represents the disease that the host of the sample is presenting with. This variable has approximately 68,000 missing values. For values that are not missing, the values are some variation of "Campylobacteriosis", "Gastroenteritis", or "Campylobacter gastroenteritis", which would all be the same disease in a clinical setting. As a result, this variable may not may not be very useful in our analysis so we may not keep it in our final dataset.

* **Outbreak**
```{r, eval=FALSE}
# display structure of host disease
str(data$Outbreak)
# generate sum of missing value counts in host disease
sum(is.na(data$Outbreak))
```

"Outbreak" is a categorical character variable that represents the identifier for an outbreak that an isolate is tied to. We have 80,107 missing values for this variable, meaning we only have 7 observed values. As a result, this variable will be pretty useless to our analysis and we will have to create our own Outbreak variable based on what we have learned.
