---
title: "Variables Description" 
author: "Anusha Kumar, Kyla Finlayson, Nataliya Kyrychenko"
date: "2022-10-18"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

### **Data Description**

In this study, secondary data is used for Campylobacter jejuni, taken from the "Isolates" browser about pathogen detection in the National Library of Medicine (NIH).

Since this data is isolated within a database, multiple people have collected the data and updated it. The database-level attribute known as isolation governs when and how modifications are made as well as whether they are made available to other individuals/systems.  Allowing numerous transactions to take place at once, without negatively influencing the execution of each, is one of the objectives of isolation. As a result, it remains challenging to determine how this data was collected as further information is not provided regarding how the individuals received the data.

The specific data of interest for our project, Campylobacter jejuni, contains 80095 rows or observations with 16 columns or variables and can be located at https://www.ncbi.nlm.nih.gov/pathogens/isolates/#taxgroup_name:%22Campylobacter%20jejuni%22. 

### **Variable Descriptions**

```{r}
# installing libraries
library(readr) # assists with reading rectangular data (csv)
library(kableExtra) # build tables and manipulate table styles
library(dplyr) # for data manipulation
library(ggplot2) # for creating plots
```

```{r}
# Loading in C. jejuni isolates data from NIH
# renaming it as "data"
data<-read.csv("/Users/anushakumar/Documents/Grad School Documents (Macbook)/PHP 2550/Project/isolates.csv")
# converting the data to data frame
data<-as.data.frame(data)
# examining first 10 rows
# and first 4 columns
# creating a table
head(data[,1:4],10)%>%
  kbl(caption = "First 10 rows and first 4 columns in the data") %>%
  kable_classic(full_width = F, html_font = "Cambria")
```

* **Organism group**
```{r}
# structure of data by organism group
str(data$Organism.group)
# sum of missing value counts in organism group
sum(is.na(data$Organism.group))
```

The organism group is a character variable, with only one level namely "Campylobacter jejuni." In the developed world, the prevalence of food borne infections, which arise from Campylobacter jejuni, can be comparable or is even higher in quantity than that of Salmonella (Source: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5584092/). Although it is rare that humans can be infected from animal and environmental reservoirs, chickens are the main cause of concern (and main culprit). 

The organism group shows the group in which an organism is classified into by taxonomy (in this case, we have all campylobacter jejuni). There are no missing values indicated within this variable.

*  **Strain**

```{r}
# structure of data by strain
str(data$Strain)
# sum of missing value counts in strain
sum(is.na(data$Strain))
```

The strain is a character variable. A strain can be described as a subtype or genetic variation of a microorganism (e.g., a virus, bacterium or fungus, in other words, a microbial strain name, if provided by the person who submits this information). For example, when we refer to a "flu strain," this is a specific biological variation of the influenza virus. 

There is no missing values within our strain variable.

* **Isolate identifiers**
```{r}
# structure of data by isolate identifiers
str(data$Isolate.identifiers)
# sum of missing value counts in isolate identifiers
sum(is.na(data$Isolate.identifiers))
```

The isolate identifiers is a character variable with no any missing values. It shows the identity of each isolate, and can contain special characters (hyphens). Isolation is the process of removing a strain from a naturally occurring, diverse population of living bacteria that is present in the environment (Source: https://www.tmcc.edu/microbiology-resource-center/lab-protocols/bacterial-isolation).

* **Serovar**
```{r}
# display structure of serovar variable 
str(data$Serovar)
# generate serovar counts
# name as summary
Summary<-data %>% count(Serovar)
# generate names of summary
# serovar and count columns
names(Summary)<-c("Serovar","Count")
# add percentage column
# create into a table
Summary$Percentage<-Summary$Count/sum(Summary$Count)*100
head(Summary,15)%>%
  kbl(caption = "Distribution of Serovar") %>%
  kable_classic(full_width = F, html_font = "Cambria")
```

The serovar is a character categorical variable with ten levels and several missing values. It represents the combined field of sub-species, serotype, or serovar, if provided. There are 79663 unrecorded serovar values with the largest type being unknown, which has has 377 observations that is equal to appropriately 0.4706911%.

* **Isolate**

```{r}
# display structure of isolate
str(data$Isolate)
# display total count of isolate missing values
sum(is.na(data$Isolate))
```

This shows separate microorganisms based on microorganism's culture or subpopulation from its parent population. The isolation variable demonstrates how to carry out this separation and control, for instance, by isolating a pathogen from sick plant tissue. This is a character variable with no missing values.

* **Create date**
```{r}
# display structure of create date
str(data$Create.date)
# generate sum of missing value counts in create date
sum(is.na(data$Create.date))
```

This shows the date when the table of data set was entered or created in the database. In other words, this represents the data on which the particular isolate was seen/detected by the Pathogen Detection System. The format used is the YYYY-MM-DD format (follows ISO format). It has no missing values.

* **Location**

```{r}
# display structure of location
str(data$Location)
# display location count
# name as summary
Summary<-data %>% count(Location)
# display names of summary
# location and count columns
names(Summary)<-c("Location","Count")
# add percentage column
# form into a table
Summary$Percentage<-Summary$Count/sum(Summary$Count)*100
head(Summary,15)%>%
  kbl(caption = "Distribution of Location") %>%
  kable_classic(full_width = F, html_font = "Cambria")
```

This shows the location where the data was collected (geographic location of the sample). There are two parts to the location, indicated by the colon (:). The name before the colon represents the country, while the name after the colon represents the region. The location is a character categorical variable showing 448 locations with 4906 missing observations.

* **Isolation source**

```{r}
# display structure of isolation source
str(data$Isolation.source)
# add count of isolation source
Summary<-data %>% count(Isolation.source)
# generate names
# isolation source and count columns
names(Summary)<-c("Isolation.source","Count")
# add percentage 
# display in table
Summary$Percentage<-Summary$Count/sum(Summary$Count)*100
head(Summary,15)%>%
  kbl(caption = "Distribution of Isolation.source") %>%
  kable_classic(full_width = F, html_font = "Cambria")
```
This shows the source of each isolation (where isolation was based on). For instance, the source of isolation can be stool or chicken carcass among many others. More particularly, the isolation source describes the physical, environmental, and/or local geographical source. The variable is a categorical in nature with character values. It records 658 different isolation sources with 18229 missing values.

* **Isolation type**

```{r}
# display structure of isolation type
str(data$Isolation.type)
# add count of isolation type
Summary<-data %>% count(Isolation.type)
# generate names
# isolation type and count columns
names(Summary)<-c("Isolation.type","Count")
# add percentage column
# display in table 
Summary$Percentage<-Summary$Count/sum(Summary$Count)*100
# create summary with distribution of isolation type
Summary%>%
  kbl(caption = "Distribution of Isolation.type") %>%
  kable_classic(full_width = F, html_font = "Cambria")
# create plot
# percentage of isolation types 
ggplot(Summary, aes(x = Isolation.type, y=Percentage,fill=Isolation.type)) +
  geom_bar(stat="identity",position=position_dodge(),width = 0.98)+theme_minimal()+ ggtitle("") +
  xlab("Isolation.type") + ylab("Percentage")+
theme(axis.text.x=element_blank(),
      axis.ticks.x=element_blank(),
      )
```

This shows the type of isolation in each observation (clinical OR environmental/other OR NULL). This is a character categorical variable with only two levels and 3260 missing values, which is equivalent to 4.070167%. Environmental/other isolation types appears most in the data set, with 43282 recordings, which is approximately 54.038330%.

* **SNP cluster**

```{r}
# structure of SNP cluster
str(data$SNP.cluster)
# sum of missing value counts in SNP cluster
sum(is.na(data$SNP.cluster))
```

The DNA sequence variations are known as single nucleotide polymorphisms (SNPs). They occur when a single nucleotide in the genome varies between matched chromosomes (Source: https://medlineplus.gov/genetics/understanding/genomicresearch/snp/#:~:text=Single%20nucleotide%20polymorphisms%2C%20frequently%20called,building%20block%2C%20called%20a%20nucleotide.). This variable has character values without any missing values.

* **BioSample**

```{r}
# display structure of biosample
str(data$BioSample)
# display total count of biosample missing values
sum(is.na(data$BioSample))
```

This shows the biological sample. This variable is a character with no missing value.


* **AMR genotypes**
```{r}
# structure of AMR genotypes
str(data$AMR.genotypes)
# total count of AMR genotypes missing values
sum(is.na(data$AMR.genotypes))
```

This shows AMR genotypes in each observation. The genes linked to or implicated in the resistance to one or more antibiotics are referred to as AMR genes. The existence or absence of a gene, or the acquisition of particular mutations accidentally or over the course of the evolution, may be the factor for resistance (Source: https://www.who.int/news-room/fact-sheets/detail/antimicrobial-resistance). It is a character variable without any missing values.

More specifically, the SNP cluster is a group of isolates, where the genome assemblies are closely linked to each other. This is dependent on the type of clustering methodology.

* **Computed types**

```{r}
# structure of computed types
str(data$Computed.types)
# display levels of computed types
levels(factor(data$Computed.types))
```

This shows the type of each isolate as per the computational pathology definitions. Unfortunately, there is no single entry in this variable.

* **Min.same**
```{r}
# display structure of min same
str(data$Min.same)
# histogram of min same
hist(data$Min.same)
# change min same to factor
data$same<-as.factor(data$Min.same)
# distribution of min same
# count and percentage columns
# create table
Summary<-data %>% count(same)
names(Summary)<-c("Min.same","Count")
Summary$Percentage<-Summary$Count/sum(Summary$Count)*100
head(Summary,15)%>%arrange(-Count)%>%
  kbl(caption = "Distribution of Min.same") %>%
  kable_classic(full_width = F, html_font = "Cambria")
```

This variable represents the minimum SNP distance from this isolate to one of the same isolation type. The distribution of this variable seems to be skewed right, indicating a violation of the normality distribution. From analyzing the distribution of each recorded number, a distance of zero is the most common with 10842 counts, which is approximately 13.5364255%. There are 33390 missing values, which is equivalent to 41.6879955%.

* **Min.diff**
```{r}
# structure of min diff
str(data$Min.diff)
# histogram of min diff
hist(data$Min.diff)
# change min diff to factor
data$diff<-as.factor(data$Min.diff)
# generate distribution table
Summary<-data %>% count(diff)
names(Summary)<-c("Min.diff","Count")
Summary$Percentage<-Summary$Count/sum(Summary$Count)*100
head(Summary,15)%>%
  kbl(caption = "Distribution of Min.diff") %>%
  kable_classic(full_width = F, html_font = "Cambria")
```

This is an integer variable representing the minimum SNP distance from the isolate to one of a different isolation type. The distribution of this variable seems to be skewed right, indicating a violation of the normality distribution. From analyzing the distribution of each recorded number, a distance of two is the most common with 3240 counts, which is approximately 4.0451963%. There are 33390 missing values, which is equivalent to 41.6879955%.
